### DISCUSSION TITLE
Need some Help with SSD1963 and RP2040

### DISCUSSION THREAD

**Discussion by Technikfreak2002** [Bodmer/TFT_eSPI#2283] [Compatible displays and setup files]
Hello! 
First of all, thank your awesome library!
I have some trouble with a SSD1963 Display. It works fine with an ESP32, 8 bit mode, SSD1963_DIVERALT, and RGB order.
But if i use another MCU, in this case a Pi Pico, i get only wired colors and pixel garbage on the screen.
I also tried some STM32 boards, with no luck. STM32F103, STM32F411 (Blue and Black Pill) Nucleo64 L476RG and Nucleo144 H743ZI2. 
for every test i made a user setup to set the correct pins. but no luck.
Other display drivers, like the ILI9486, work with the pico and at with the nucleo64.

Can some give me a hint? am i missing a bit?

Thank you very much and a happy new year! :D

**Comment by Bodmer** [comment]
Post your setup file for the Pico. Try a slower CPU clock speed.

**Comment by Technikfreak2002** [comment]
Hello Bodmer!
Thank you for your reply!
I used Setup_103 and added SSD1963 driver for use with the pico. For the esp32 Setup_50 unmodified.
Clock speed was sefault 133MHz but i also tested down to 50MHz. 
I use arduino IDE 1.8.19 with Pico SDK from earlephilhower.
For reference, the Pico / display combination works fine with PicoMite. Different pins, but same display and same wires.

I make a small PCB today to keep the wires as short as possible and give it a try. 

`
// This setup is for the RP2040 processor only when used with 8 bit parallel displays
// See SetupX_Template.h for all options available
#define USER_SETUP_ID 103

////////////////////////////////////////////////////////////////////////////////////////////
// Interface
////////////////////////////////////////////////////////////////////////////////////////////
#define TFT_PARALLEL_8_BIT

////////////////////////////////////////////////////////////////////////////////////////////
// Display driver type
////////////////////////////////////////////////////////////////////////////////////////////
//#define ILI9486_DRIVER
//#define SSD1963_480_DRIVER    // 272 x 480 display
//#define SSD1963_800_DRIVER    // 480 x 800 display
#define SSD1963_800ALT_DRIVER // Alternative 480 x 800 display
//#define SSD1963_800BD_DRIVER    // 480 x 800 display sourced from https://www.buydisplay.com/7-tft-screen-touch-lcd-display-module-w-ssd1963-controller-board-mcu

#define TFT_RGB_ORDER TFT_RGB  // Colour order Red-Green-Blue
//#define TFT_RGB_ORDER TFT_BGR  // Colour order Blue-Green-Red

////////////////////////////////////////////////////////////////////////////////////////////
// RP2040 pins used
////////////////////////////////////////////////////////////////////////////////////////////

//#define TFT_CS   -1  // Do not define, chip select control pin permanently connected to 0V

// These pins can be moved and are controlled directly by the library software
#define TFT_DC   28    // Data Command control pin
#define TFT_RST  27    // Reset pin

//#define TFT_RD   -1  // Do not define, read pin permanently connected to 3V3

// Note: All the following pins are PIO hardware configured and driven
// The pins are hard-coded at the moment and must not be changed here
// Connections MUST use the pins below
  #define TFT_WR   22

  // PIO requires these to be sequentially increasing - do not change
  #define TFT_D0    6
  #define TFT_D1    7
  #define TFT_D2    8
  #define TFT_D3    9
  #define TFT_D4   10
  #define TFT_D5   11
  #define TFT_D6   12
  #define TFT_D7   13
//*/

////////////////////////////////////////////////////////////////////////////////////////////
// Fonts to be available
////////////////////////////////////////////////////////////////////////////////////////////
#define LOAD_GLCD   // Font 1. Original Adafruit 8 pixel font needs ~1820 bytes in FLASH
#define LOAD_FONT2  // Font 2. Small 16 pixel high font, needs ~3534 bytes in FLASH, 96 characters
#define LOAD_FONT4  // Font 4. Medium 26 pixel high font, needs ~5848 bytes in FLASH, 96 characters
#define LOAD_FONT6  // Font 6. Large 48 pixel font, needs ~2666 bytes in FLASH, only characters 1234567890:-.apm
#define LOAD_FONT7  // Font 7. 7 segment 48 pixel font, needs ~2438 bytes in FLASH, only characters 1234567890:.
#define LOAD_FONT8  // Font 8. Large 75 pixel font needs ~3256 bytes in FLASH, only characters 1234567890:-.
#define LOAD_GFXFF  // FreeFonts. Include access to the 48 Adafruit_GFX free fonts FF1 to FF48 and custom fonts

#define SMOOTH_FONT

////////////////////////////////////////////////////////////////////////////////////////////
`

**Reply by Bodmer** [reply]
Which version of TFT_eSPI are you using?

**Comment by Technikfreak2002** [comment]
I'm using 2.4.79. 
I made some tests with a small breadboard PCB and wires as short as possible. The SSD1963 does not work and a second Display with ILI9341 8Bit also does not work.
To make sure i wired all correct i tested all with mcufriend_kvb. And all works. Here is the pin definition code inside mcfriend_special.h
`
//################################### RP2040 ##############################
#elif defined(USE_MY_PICO) && defined(ARDUINO_ARCH_RP2040)       //regular UNO shield on PICO
//LCD pins  |D7  |D6  |D5  |D4  |D3  |D2 |D1  |D0  | |RD  |WR  |RS  |CS  |RST |    |
//RP2040 pin|GP13|GP12|GP11|GP10|GP9 |GP8|GP19|GP18| |GP14|GP26|GP27|GP28|GP16|GP17|
//BODMER pin|GP13|GP12|GP11|GP10|GP9 |GP8|GP7 |GP6 | |GP14|GP22|GP27|GP28|GP16|GP17|
//UNO pins  |7   |6   |5   |4   |3   |2  |9   |8   | |A0  |A1  |A2  |A3  |A4  |A5  |
//LCD pins  |CS  |MOSI|MISO|SCK | |SDA|SCL|
//RP2040 pin|GP21|GP3 |GP4 |GP2 | |GP6|GP7|
//UNO pins  |10  |11  |12  |13  | |18 |19 |

#define USE_BODMER

#define WRITE_DELAY { WR_ACTIVE4; }
#define IDLE_DELAY  { WR_IDLE; }
#define READ_DELAY  { RD_ACTIVE8; }

#define RD_PORT sio_hw
#define RD_PIN  14
#define WR_PORT sio_hw
#define WR_PIN  26
#define CD_PORT sio_hw
#define CD_PIN  27
#define CS_PORT sio_hw
#define CS_PIN  28
#define RESET_PORT sio_hw
#define RESET_PIN  16

#define LMASK         (0x03 << 18)
#define HMASK         (0xFC << 6)
#if defined(USE_BODMER)
#define WR_PIN        22
#define CD_PIN        28
#define CS_PIN        16
#define RESET_PIN     27
#define LMASK         (0x00 << 18)
#define HMASK         (0xFF << 6)
#endif
#define GPMASK        (LMASK | HMASK)       //more intuitive style for mixed Ports
#define write_8(x)    { sio_hw->gpio_clr = GPMASK; \
                        sio_hw->gpio_set = (((x) << 18) & LMASK) | (((x) << 6) & HMASK); \
                      }
#define read_8()      ( ((sio_hw->gpio_in & HMASK) >> 6) | ((sio_hw->gpio_in & LMASK) >> 18) )
#define setWriteDir() { sio_hw->gpio_oe_set = GPMASK; }
#define setReadDir()  { sio_hw->gpio_oe_clr = GPMASK; }
#define GPIO_INIT() {for (int i = 6; i <= 14; i++) pinMode(i, OUTPUT); for (int i = 16; i <= 28; i++) pinMode(i, OUTPUT);}
#define write8(x)     { write_8(x); WRITE_DELAY; WR_STROBE; IDLE_DELAY; }
#define write16(x)    { uint8_t h = (x)>>8, l = x; write8(h); write8(l); }
#define READ_8(dst)   { RD_STROBE; READ_DELAY; dst = read_8(); RD_IDLE2; RD_IDLE; }
#define READ_16(dst)  { uint8_t hi; READ_8(hi); READ_8(dst); dst |= (hi << 8); }

#define PIN_LOW(port, pin)    (port)->gpio_clr = (1<<(pin))
#define PIN_HIGH(port, pin)   (port)->gpio_set = (1<<(pin))
#define PIN_OUTPUT(port, pin) (port)->gpio_oe_set = (1<<(pin))
`

**Comment by Bodmer** [comment]
Have you wired  CS pin on display to GND, and RD pin on display to 3.3V when using TFT_eSPI?

**Reply by Technikfreak2002** [reply]
Yes. Also with mcufriend.

**Comment by Bodmer** [comment]
I have raised an issue as it should work with your setup. Thanks for reporting this.

**Comment by Technikfreak2002** [comment]
Thank You very much! I will follow the progress!

**Comment by Bodmer** [comment]
Fixed in master library. Thanks for repporting this.

**Comment by Technikfreak2002** [comment]
Hello! Sorry for my late reply!
I tested the new version. And it works! Thank you so much!
